### PR TITLE
Feature/track raw score

### PR DESCRIPTION
**Pull request by isabella618033**

### CODE REVIEW

File: prompting/validators/reward/dahoas.py [hunk 1]
```diff
@@ -92,21 +97,27 @@ def reward_fn(samples):
...
-            dtype=torch.float32,
-        ).to(self.device)
+    ) -> dict:
+        # Get all the reward results.
+        reward_events = [self.reward(prompt, completion, name) for completion in completions]
+
+        # Parse the result and generate an event to be logged.
+        parsed_reward_events = BaseRewardEvent.parse_reward_events(reward_events)
+
+        parsed_reward_events['reward'] = torch.tensor(parsed_reward_events['reward'], dtype=torch.float32).to(self.device)
```
Comment (p-ferreira) [inline prompting/validators/reward/dahoas.py:118]: What if `parse_reward_events` handles this float conversion? It seems that we are repeating this code across the reward models

File: prompting/validators/forward.py [hunk 2]
```diff
@@ -122,19 +122,17 @@ async def run_step(
         self.device
     )
     for weight_i, reward_fn_i in zip(self.reward_weights, self.reward_functions):
-        reward_i, reward_i_normalized = reward_fn_i.apply(prompt, responses, name)
+        reward_i_normalized, reward_event = reward_fn_i.apply(prompt, responses, name)
```
Comment (p-ferreira) [inline prompting/validators/forward.py:125]: What if we get the reward event that contains a normalized value within it? That way the flow would be simplified around a single entity `reward_event`

File: prompting/validators/reward/diversity.py [hunk 3]
```diff
@@ -173,9 +174,9 @@ def get_rewards(
 
         # Return all
         if historic_rewards != None:
-            return batch_rewards * historic_rewards
+            return {'reward': batch_rewards * historic_rewards}
```
Comment (p-ferreira) [inline prompting/validators/reward/diversity.py:177]: Are we returning a raw dict here? What if we use the `BaseRewardModel` to streamline the behavior of get_rewards ->  `BaseRewardEvent` or any derivative class

File: prompting/validators/reward/nsfw.py [hunk 4]
```diff
@@ -63,15 +69,23 @@ def sum_nsfw_scores(input_ids, chunk_size):
                 return max_score
 
             # 0 when needs to be filtered out, 1 when it is safe
-            return 0.0 if sum_nsfw_scores(input_ids, chunk_size=512) > boundary else 1.0
+            nsfw_score = sum_nsfw_scores(input_ids, chunk_size=512)
+            reward_event.nsfw_score = nsfw_score
```
Comment (p-ferreira) [inline prompting/validators/reward/nsfw.py]: why are we using `nsfw_score` instead of the default `normalized_score` that we were using before?

**Review by p-ferreira — COMMENTED** [review]
What if we streamline the whole process into an abstraction where:
- `get_rewards` return a `BaseRewardEvent`,
- `BaseRewardModel.apply` returns a `BaseRewardEvent`
- `BaseRewardEvent` has the logic to convert it itself into a dictionary, from which we can spread its properties later in the `forward` call and in the event schema

File: prompting/validators/reward/nsfw.py (hunk 4 above)
Comment (isabella618033) [reply prompting/validators/reward/nsfw.py]: so the nsfw_score would be the raw score 
nsfw_reward = nsfw_normalized would be the (0/1) after passing the boundary

File: prompting/validators/reward/dahoas.py (hunk 1 above)
Comment (isabella618033) [reply prompting/validators/reward/dahoas.py]: Hmm, I would prefer keeping it here cause how to post-process the reward events (eg. parsing into a tensor) seems to be the task of the reward model itself.

File: prompting/validators/mock.py [hunk 5]
```diff
@@ -59,7 +59,7 @@ def __init__(self, mock_name: str = "MockReward"):
 
     def apply(self, prompt: str, completion: List[str], name: str) -> torch.FloatTensor:
         mock_reward = torch.tensor([1 for _ in completion], dtype=torch.float32)
-        return mock_reward, mock_reward
+        return mock_reward, {}
```
Comment (p-ferreira) [inline prompting/validators/mock.py:62]: Good catch!

File: prompting/validators/reward/diversity.py (hunk 3 above)
Comment (isabella618033) [reply prompting/validators/reward/diversity.py]: just fixed! thank you:)

**Review by p-ferreira — APPROVED** [review]
LGTM (BTW I set the PR as ready for review by accident, apologies for that) 

One thing missing though:
- [ ] Fixing branch conflicts

File: prompting/validators/reward/dpo.py [hunk 6]
```diff
@@ -51,15 +51,21 @@ def __init__(self, device: str):
...
         Uses guidance from https://github.com/eric-mitchell/direct-preference-optimization/blob/main/trainers.py.
         """
+
+        reward_event = BaseRewardEvent()
+
         with torch.no_grad():
             # Check if completion is
             if completion.strip() == "" or len(completion) <= 5:
-                return -11  # exp(-11)=1.67e-5 < 2e-5=1/50257 (typical vocab size)
+                reward_event.reward = (
```
Comment (steffencruz) [inline prompting/validators/reward/dpo.py:65]:
```suggestion
                # exp(-11)=1.67e-5 < 2e-5=1/50257 (typical vocab size)
                reward_event.reward = -11.0
```

File: prompting/validators/reward/dpo.py [hunk 7]
```diff
@@ -51,15 +51,21 @@ def __init__(self, device: str):
...
+                    -11.0
+                )  # exp(-11)=1.67e-5 < 2e-5=1/50257 (typical vocab size)
```
Comment (steffencruz) [inline prompting/validators/reward/dpo.py:67]:
```suggestion
               
```

File: prompting/validators/reward/dpo.py [hunk 8]
```diff
@@ -74,7 +80,10 @@ def reward_single(
 
             # Completion doesn't fit into model sequence, so return lowest reward.
             if self.tokenizer.model_max_length <= len(prompt_part):
-                return -11.0  # exp(-11)=1.67e-5 < 2e-5=1/50257 (typical vocab size)
+                reward_event.reward = -11.0
+                return (
```
Comment (steffencruz) [inline prompting/validators/reward/dpo.py:84]:
```suggestion
                return reward_event
```

File: prompting/validators/reward/dpo.py [hunk 9]
```diff
@@ -74,7 +80,10 @@ def reward_single(
...
+                    reward_event  # exp(-11)=1.67e-5 < 2e-5=1/50257 (typical vocab size)
```
Comment (steffencruz) [inline prompting/validators/reward/dpo.py:85]:
```suggestion
                  
```

File: prompting/validators/reward/dpo.py [hunk 10]
```diff
@@ -74,7 +80,10 @@ def reward_single(
...
+                )
```
Comment (steffencruz) [inline prompting/validators/reward/dpo.py:86]:
```suggestion
                
```

File: prompting/validators/reward/dpo.py [hunk 11]
```diff
@@ -123,18 +132,21 @@ def reward_single(
 
             # NaNs can possibly arise through log(0)=-inf, replace with suitably small logits.
             if torch.isnan(reward) or torch.isinf(reward):
-                return -11.0  # exp(-11)=1.67e-5 < 2e-5=1/50257 (typical vocab size)
-            return reward.item()
+                reward_event.reward = (
```
Comment (steffencruz) [inline prompting/validators/reward/dpo.py:135]:
```suggestion
                reward_event.reward = 11
```

File: prompting/validators/reward/dpo.py [hunk 12]
```diff
@@ -123,18 +132,21 @@ def reward_single(
...
+                    -11.0
```
Comment (steffencruz) [inline prompting/validators/reward/dpo.py:136]:
```suggestion
                    
```

File: prompting/validators/reward/dpo.py [hunk 13]
```diff
@@ -123,18 +132,21 @@ def reward_single(
...
+                )  # exp(-11)=1.67e-5 < 2e-5=1/50257 (typical vocab size)
```
Comment (steffencruz) [inline prompting/validators/reward/dpo.py:137]:
```suggestion
            
```

File: prompting/validators/reward/open_assistant.py [hunk 14]
```diff
@@ -40,20 +40,22 @@ def __init__(self, device: str):
...
-    def reward_single(self, prompt: str, completion: str, name: str) -> float:
+    def reward_single(self, prompt: str, completion: str, name: str) -> BaseRewardEvent:
+        reward_event = BaseRewardEvent()
+
         with torch.no_grad():
             inputs = self.tokenizer(prompt, completion, return_tensors="pt").to(
                 self.device
             )
-            return float(self.model(**inputs).logits[0].cpu().detach())
+            reward_event.reward = float(self.model(**inputs).logits[0].cpu().detach())
```
Comment (steffencruz) [inline prompting/validators/reward/open_assistant.py:50]: shouldn't this be `reward_event.score` since its float? where is `reward_event.reward:Bool`?

Comment (isabella618033) [reply prompting/validators/reward/open_assistant.py:50]: rlhf is a reward model (not filter model), so there is no bool reward, and the score = reward
